### ISSUE TITLE
IDE0004 doesn't trigger for method arguments to local function in same scope

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#38347]
**Version Used**: 
3.3.0-beta3-19415-01+f5ba9f2c61a2fe853dc4913888d40df221539147

**Steps to Reproduce**:
```C#
class Program
{
    static void Main() { }

    public static void M()
    {
        for (int i = 0; i < 1; i++)
        {
            long a = 0, b = 0;

            decimal result = (decimal)a + (decimal)b; // IDE004
            SameScope(result);

            OuterScope((decimal)a + (decimal)b); // IDE004

            SameScope((decimal)a + (decimal)b); // ## Not offered ##

            static void SameScope(decimal sum) { }
        }

        static void OuterScope(decimal sum) { }
    }
}
```

**Expected Behavior**:
IDE004 offered on all casts to decimal.

**Actual Behavior**:
IDE004 isn't offered on the casts in `SameScope((decimal)a + (decimal)b)`.
![image](https://user-images.githubusercontent.com/2642209/63890672-7886b800-c9b1-11e9-9e2a-5fb6601498d6.png)

**Comment by RikkiGibson** [comment]
@jasonmalinowski can you remind me what mechanism the IDE uses to decide if a cast is redundant?

**Comment by jasonmalinowski** [comment]
The algorithm is...complicated. The function entrypoint is this:

https://github.com/dotnet/roslyn/blob/7bc44488c661fd6bbb6c53f39512a6fe0cc5ef84/src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs#L321

Which as you may notice is a 267 line function, which calls a number of helpers and uses a number of fairly complicated additional helpers. :smile:

**Comment by jasonmalinowski** [comment]
But to be more serious: high level it's a combination of calls to ClassifyConversion along with a speculative binding pass that ensures that removing a cast doesn't also change overload resolution.
